mod_clamav.c

```diff
@@ -77,70 +77,74 @@ static int clamavd_result(int sockd, const char *abs_filename, const char *rel_f
   }
 
   memset(buff, '\0', sizeof(buff));
-  if (fgets(buff, sizeof(buff) - 1, fd)) {
-    if (strstr(buff, "FOUND\n")) {
-      const char *proto;
-
-      ++infected;
-
-      pt = strrchr(buff, ':');
-      if (pt)
-        *pt = 0;
-
-      /* Delete the infected upload */
-      /* TODO: delete only the hiddenstore file, not
-         the actual file.
-      */
-      if ((ret=pr_fsio_unlink(rel_filename))!=0) {
-        pr_log_pri(PR_LOG_ERR,
-                   MOD_CLAMAV_VERSION ": notice: unlink() failed (%d): %s",
-                   errno, strerror(errno));
-      }
+  // Try to read in a loop in case fgets gets interrupted
+  do {
+    errno = 0;
+    fgets(buff, sizeof(buff) - 1, fd);
+    fxerrno = errno;
+  } while (fxerrno == EINTR);
```

Comment:
You might want to include handling of the interrupting signal in this loop, _e.g._:
```
do {
  errno = 0;
  pr_signals_handle();
  fgets(buff, sizeof(buff) - 1, fd);
  fxerrno = errno;
} while (fxerrno == EINTR);
```

Reply:
OK, I'm not super familiar with that particular function, but I see no issue with it.  The main signal that I have run into so far is SIGALRM, which fires every 5 seconds (or less), and scanning of larger archives can take more than that (the SIGALRM was interrupting the fgets, which was not resumed so that the response was never received from clamd).  Is there not a signal handler that fires right away?

Reply:
I added pr_dignals_handle() to my patch.  I also didn't realize that my previous patch had been merged already, which is where the conflict comes from.  So I need to figure out how to fix that too...

Reply:
ProFTPD redid its signal handlers many years ago, so that they do not run asynchronously.  The state of the stack isn't always reliable/predictable when code is running in a signal handler, and there have been security issues in C code in such cases.

So instead, ProFTPD signal handlers simply set a bitflag, and exit.  The actual _processing_ of that code that should run is done later, via `pr_signals_handle()`.  This way, the calling function can decide _when_ it wants the actual signal-handling code (such as timers) to be done; maybe it should be done after the calling code finishes a tight loop, for example.  Or maybe not.

Most loops in ProFTPD have a call to `pr_signals_handle()` within the loop, especially if there's network/disk IO involved.  Such IO can be blocking, and if the admin wants to terminate the daemon when the daemon is in such a loop -- and there wasn't a call to `pr_signals_handle` in the loop -- then the process would sit there still running, until the admin became impatient, and used `SIGKILL` (`kill -9`), which is not catchable.